clippy_lints/src/utils/internal_lints/metadata_collector.rs

```diff
@@ -263,14 +269,192 @@ impl LintMetadata {
     }
 }
 
+fn replace_produces(lint_name: &str, docs: &mut String, clippy_project_root: &Path) {
+    let mut doc_lines = docs.lines().map(ToString::to_string).collect::<Vec<_>>();
+    let mut lines = doc_lines.iter_mut();
+
+    'outer: loop {
+        // Find the start of the example
+
+        // ```rust
+        loop {
+            match lines.next() {
+                Some(line) if line.trim_start().starts_with("```rust") => {
+                    if line.contains("ignore") || line.contains("no_run") {
+                        // A {{produces}} marker may have been put on a ignored code block by mistake,
+                        // just seek to the end of the code block and continue checking.
+                        if lines.any(|line| line.trim_start().starts_with("```")) {
+                            continue;
+                        }
+
+                        panic!("lint `{}` has an unterminated code block", lint_name)
+                    }
+
+                    break;
+                },
+                Some(line) if line.trim_start() == "{{produces}}" => {
+                    panic!(
+                        "lint `{}` has marker {{{{produces}}}} with an ignored or missing code block",
+                        lint_name
+                    )
+                },
+                Some(line) => {
+                    let line = line.trim();
+                    // These are the two most common markers of the corrections section
+                    if line.eq_ignore_ascii_case("Use instead:") || line.eq_ignore_ascii_case("Could be written as:") {
+                        break 'outer;
+                    }
```

Comment:
Yep, this is just an attempt to do as little work as possible. Normally when lints have multiple examples they use "or" anyway, so it shouldn't be an issue.

Reply:
Okay, I'm debating if this actually has an huge impact on the runtime. :thinking:

I also want to see how what effect this PR has on the runtime. But I doubt that we can really do any optimization here, since we need to compile every snippet separately. Therefore, it should be fine.

Reply:
Here are a few timings:

<details>
<summary>Before the PR</summary>

<pre>
Executed in    8.75 secs    fish           external
   usr time   15.40 secs  310.00 micros   15.40 secs
   sys time    0.62 secs  185.00 micros    0.62 secs
</pre>
</details>

<details>
<summary>After the PR (no generated output)</summary>

<pre>
Executed in    9.44 secs    fish           external
   usr time   16.87 secs  551.00 micros   16.87 secs
   sys time    0.62 secs    0.00 micros    0.62 secs
</pre>
</details>

<details>
<summary>Generating the output of 5 lints</summary>

<pre>
Executed in   30.93 secs    fish           external
   usr time   59.21 secs    0.00 micros   59.21 secs
   sys time    3.75 secs  554.00 micros    3.75 secs
</pre>
</details>

Reply:
That's quite a step up from the original run time, and just for 5 lints. Have you looked into paralyzing this step somehow? Maybe we could start the linting processes and not wait for its output?

If it's not possible, I'm also fine with this, as it will only be run, during for merge commits.

Before we merge this, I want to try adding this to the bors workflow, to ensure that the metadata collection always succeeds. Currently, it's only run on master after the merge and can therefore fail without us noticing. I have a good idea how this can be done, and I'll create a PR on the weekend, unless you want to give it a try :upside_down_face:

Reply:
With rayon I was able to generate the output of 20 lints in about the same time:
<pre>
Executed in   29.66 secs    fish           external
   usr time   61.64 secs  430.00 micros   61.64 secs
   sys time    4.83 secs    0.00 micros    4.83 secs
</pre>

I wasn't able to get lower numbers of lints any faster though.

How do you feel about adding rayon?

Also, I'll just wait for your PR. Workflows aren't my forte :smile:.